---
params:
  path:
    value: .
  lib:
    value: x
  exp:
    value: x
  cellcon:
    value: x
  
title: '`r paste("iDrop-RNAseq Report:", params$lib)`'
author: "CellOmics Lab"
date: '`r strftime(Sys.time(), format = "%m, %d, %Y")`'
output:
  rmarkdown::html_document:
    theme: lumen
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r library,echo=FALSE}
suppressPackageStartupMessages({
 library(ggplot2)
 library(gridExtra)
 library(grid)
  library(Seurat)
      library(data.table)
})


```


```{r chunk1,fig.height=4,fig.width=12}
bc <- fread(paste(params$path, "/temp/cell_stat.txt", sep=""),header=TRUE)
bc <- as.data.frame(bc)
bc <- subset(bc, bc$nUMI>0)
bc <- bc[order(bc$nUMI, decreasing=T),]
len <- nrow(bc)
len<-length(bc$nUMI)
a = log10(1:len)
b = log10(as.numeric(bc$nUMI))
lo <- loess(b~a,span = 0.004,degree = 2)
c = log10(int(params$exp)*1.7)
if(10^c>len){c=log10(600)}
xl <- seq(2,c, (c - 2)/10)
out = predict(lo,xl)
infl <- c(FALSE,abs(diff(out)/((c - 2)/10) - -1) == min(abs(diff(out)/((c - 2)/10)- -1)))
m = 10 ^ out[infl] + 0.5
m = round(m , digits =0 )
cutoff<-length(which(bc$nUMI>=m))
tmp<-data.frame(x=1:len,y=bc$nUMI,g=c(rep("Cell",cutoff),rep("Background",len-cutoff)))


anno1.x=round(len*0.4)
anno1.y=round(max(tmp$y)*0.6)
anno2.x=round(len*0.4)
anno2.y=round(max(tmp$y)*1.2)
p =    ggplot(tmp,aes(x=x,y=y))
p = p +geom_line(aes(color=g),size=2)
p = p +scale_x_log10(name="Barcodes")
p = p +scale_y_log10(name="UMI counts",breaks=c(1,10,100,1000,10000,100000),labels=c(1,10,100,"1k","10K","100K"))
p = p +annotate("text",x=1,y=1,label=paste("Cell =",cutoff),size=6,hjust=0)
p = p +annotate("text",x=1,y=2.5,label=paste("UMI =",m),size=6,hjust=0) + labs(color="")

cc <- bc[1:cutoff,]
called = sum(cc$nUMI)
in_cell <- round(called/sum(bc$nUMI),digits=4)


small<-bc[which(bc$nUMI>=m),]
CellNum = nrow(small)
UMI_mean = mean(small$nUMI)
UMI_median = median(small$nUMI)
Gene_mean = mean(small$nGene)
Gene_median = median(small$nGene)
UMI_Cell = sum(small$nUMI)


p1 =    ggplot(tmp,aes(x=x,y=y)) + xlim(0,10) + ylim(0,9)
p1 = p1 +annotate("text",x=0.2,y=9,label="Estimated Number of Cell:",size=10,hjust=0)
p1 = p1 +annotate("text",x=3,y=6.5,label=CellNum,size=20,hjust=0, colour="orange")
p1 = p1 +annotate("text",x=0.2,y=3.5,label="Reads in cell:",size=6,hjust=0)
p1 = p1 +annotate("text",x=9,y=3.5,label=round(in_cell,3),size=6,hjust=1)
p1 = p1 +annotate("segment",x = 0, xend = 10, y = 3, yend = 3,colour = "green")
p1 = p1 +annotate("text",x=0.2,y=2.5,label="Mean UMI counts per cell:",size=6,hjust=0)
p1 = p1 +annotate("text",x=9,y=2.5,label=round(UMI_mean),size=6,hjust=1)
p1 = p1 +annotate("segment", x = 0, xend = 10, y = 2, yend = 2,colour = "green")
p1 = p1 +annotate("text",x=0.2,y=1.5,label="Mean Genes per cell:",size=6,hjust=0)
p1 = p1 +annotate("text",x=9,y=1.5,label=round(Gene_mean),size=6,hjust=1)
p1 = p1 +annotate("segment",x  = 0, xend  = 10, y  = 1, yend = 1,colour = "green")
p1 = p1 +theme(axis.title=element_blank(), axis.text=element_blank(), axis.ticks=element_blank(),panel.background = element_blank(), panel.grid.major=element_blank(),panel.grid.minor = element_blank())

```


```{r chunk2, fig.height=4,fig.width=12}
grid.arrange(arrangeGrob(p1, p, widths =  c(5,5)), nrow = 1)
```

```{r  fig.height=4,fig.width=8}
cell <- fread(paste(params$path,"outs/count_mtx.tsv", sep=""),header = T)
cell <- as.data.frame(cell)
names <- as.array(as.matrix(cell[1]))
names = gsub("_", "-", names)
rownames(cell) = names
cell = cell[-1]
names = gsub("-", "_", colnames(cell))
colnames(cell) = names

z <- CreateSeuratObject(counts= cell, min.cells = 3, min.features = 200)
z[["percent.mt"]] <- PercentageFeatureSet(z, pattern = "^mt-")
z <- subset(z, subset = nFeature_RNA > 200 &  percent.mt < 3)
z <- NormalizeData(z)
z <-FindVariableFeatures(z, selection.method = "vst", nfeatures = 2000)

VlnPlot(z, features = c("nFeature_RNA", "nCount_RNA"))
```
```{r}
z <- NormalizeData(z)
z <-FindVariableFeatures(z, selection.method = "vst", nfeatures = 2000)

all.genes <- rownames(z)
z <- ScaleData(z, features = all.genes)
z <- RunPCA(z, features = VariableFeatures(object = z))
z <- FindNeighbors(z, dims = 1:10)
z <- FindClusters(z, resolution = 0.5)
z <- RunUMAP(z, dims = 1:10)
DimPlot(z, reduction = "umap")

```


